language/diem-framework/modules/Diem.move

```diff
@@ -485,97 +551,337 @@ module Diem {
         include Roles::AbortsIfNotDesignatedDealer;
-        /// Preburn is published under the DesignatedDealer account.
-        ensures exists<Preburn<CoinType>>(Signer::spec_address_of(account));
+        /// PreburnQueue is published under the DesignatedDealer account.
+        include PublishPreburnQueueAbortsIf<CoinType>;
+        include PublishPreburnQueueEnsures<CoinType>;
+        ensures exists<PreburnQueue<CoinType>>(account_addr);
 
         include Roles::AbortsIfNotTreasuryCompliance{account: tc_account};
         include AbortsIfNoCurrency<CoinType>;
         aborts_if is_synthetic_currency<CoinType>() with Errors::INVALID_ARGUMENT;
-        aborts_if exists<Preburn<CoinType>>(Signer::spec_address_of(account)) with Errors::ALREADY_PUBLISHED;
+        aborts_if exists<PreburnQueue<CoinType>>(account_addr) with Errors::ALREADY_PUBLISHED;
+        aborts_if exists<Preburn<CoinType>>(account_addr) with Errors::INVALID_STATE;
 
     }
 
     ///////////////////////////////////////////////////////////////////////////
 
+
+    /// Upgrade a designated dealer account from using a single `Preburn`
+    /// resource to using a `PreburnQueue` resource so that multiple preburn
+    /// requests can be outstanding in the same currency for a designated dealer.
+    fun upgrade_preburn<CoinType>(account: &signer)
+    acquires Preburn {
+        Roles::assert_designated_dealer(account);
+        let sender = Signer::address_of(account);
+        let preburn_exists = exists<Preburn<CoinType>>(sender);
+        let preburn_queue_exists = exists<PreburnQueue<CoinType>>(sender);
+        // The DD must already have an existing `Preburn` resource in order to be upgraded.
+        if (preburn_exists && !preburn_queue_exists) {
+            let Preburn { to_burn } = move_from<Preburn<CoinType>>(sender);
+            destroy_zero(to_burn);
+            publish_preburn_queue<CoinType>(account);
+        } else {
+            assert(preburn_queue_exists && !preburn_exists, Errors::invalid_state(EPREBURN_QUEUE))
+        };
+    }
+    spec fun upgrade_preburn {
+        let account_addr = Signer::spec_address_of(account);
+        modifies global<Preburn<CoinType>>(account_addr);
+        modifies global<PreburnQueue<CoinType>>(account_addr);
+        include UpgradePreburnAbortsIf<CoinType>;
+        include UpgradePreburnEnsures<CoinType>;
+    }
+    spec schema UpgradePreburnAbortsIf<CoinType> {
+        account: signer;
+        let account_addr = Signer::spec_address_of(account);
+        let preburn = global<Preburn<CoinType>>(account_addr);
+        let preburn_exists = exists<Preburn<CoinType>>(account_addr);
+        let preburn_queue_exists = exists<PreburnQueue<CoinType>>(account_addr);
+        aborts_if preburn_exists && preburn.to_burn.value > 0 with Errors::INVALID_ARGUMENT;
+        /// Must abort if the account doesn't have the `PreburnQueue` or
+        /// `Preburn` resource to satisfy [[H4]][PERMISSION] of `preburn_to`.
```

Comment:
what is [[H4]][PERMISSION]

Reply:
This is linking to specific permissions/roles in the DIP-2 document which is done by the documentation generator (see e.g. the comment in the spec for [this function](https://github.com/diem/diem/blob/master/language/diem-framework/modules/doc/Diem.md#function-preburn_to)